aten/src/ATen/native/native_functions.yaml

```diff
@@ -3229,9 +3229,14 @@
 
 - func: is_pinned(Tensor self, Device? device=None) -> bool
   variants: method
+  dispatch:
```

Comment:
Does everything just work when we change an op from being math-composite (`CompositeImplicitAutograd`) to being backend-specific?

My first thought was that we'd need to add an entry to derivatives.yaml, but I see some other ops that are backend-specific but do not have a derivatives.yaml entry (e.g. `aten::is_set_to`), and it looks like the autograd codegen still creates dummy autograd kernels for them.

Reply:
I checked the generated code for is_pinned and it looks fine:

```
bool is_pinned(c10::DispatchKeySet ks, const at::Tensor & self, c10::optional<at::Device> device) {
  auto& self_ = unpack(self, "self", 0);
  #ifndef NDEBUG
  c10::optional<Storage> self__storage_saved =
    self_.has_storage() ? c10::optional<Storage>(self_.storage()) : c10::nullopt;
  c10::intrusive_ptr<TensorImpl> self__impl_saved;
  if (self_.defined()) self__impl_saved = self_.getIntrusivePtr();
  #endif
  auto _tmp = ([&]() {
    at::AutoDispatchBelowADInplaceOrView guard;
    return at::redispatch::is_pinned(ks & c10::after_autograd_keyset, self_, device);
  })();
  auto result = _tmp;
  #ifndef NDEBUG
  if (self__storage_saved.has_value())
    AT_ASSERT(self__storage_saved.value().is_alias_of(self_.storage()));
  if (self__impl_saved) AT_ASSERT(self__impl_saved == self_.getIntrusivePtr());
  #endif
  TORCH_CHECK_NOT_IMPLEMENTED(!(isFwGradDefined(self)), "Trying to use forward AD with is_pinned that does not support it.");
  return result;
} 
```

These functions are weird because they don't have any meaningful autograd content (as they don't return a Tensor). Effectively, all it does exclude autograd key and then redispatch. I think this is just the AD codegen being smart enough to not error if it sees the function doesn't return a Tensor. cc @albanD for any further thoughts.

Reply:
This is exactly what happens: we only generate actual code for ops that have either one differentiable input or output: https://github.com/pytorch/pytorch/blob/b7298f499d6a88bf2da92d4f07c6f4a52e73adf4/tools/autograd/gen_variable_type.py#L461

Note that if you have differentiable input/output but you don't specify anything in derivatives.yaml, it is equivalent to doing:
```
- name: my_fn(Tensor foo, Tensor bar, int baz) -> Tensor
  foo: not_implemented("my_fn")
  bar: not_implemented("my_fn")
```

Reply:
I have a question on how this would work for a non CUDA backend. Since the empty_cpu() uses a pinned memory allocator for CUDA alone, are the other backends explicitly required to call a tensor.pin_memory() afterwards?

Reply:
Yes. And then we can fix this if we change `pin_memory` argument on empty to take a DEVICE rather than a bool, but that is a very complicated change and very BC breaking without very careful handling so you shouldn't count on it happening in the near future.